allennlp/training/trainer.py

```diff
@@ -980,11 +980,34 @@ def from_params(cls,
         grad_norm = params.pop_float("grad_norm", None)
         grad_clipping = params.pop_float("grad_clipping", None)
         lr_scheduler_params = params.pop("learning_rate_scheduler", None)
+        fine_tuning = params.pop("fine-tuning", None)
```

Comment:
If this is a boolean flag, as it looks like it is, it's better to just make it boolean, like `discriminative-fine-tuning`, unless we really think that in the near future we're going to have other possible values for the `fine-tuning` flag.

Reply:
Also, this isn't your fault, but it's concerning to me that there's no natural place to document this new parameter (c.f. @nelson-liu's find in #1635).  @joelgrus, @DeNeutoy, any good ideas on how to fix this documentation problem?

Reply:
To add: it’d be great if the key in the param name were “fine_tuning” instead of “fine-tuning” for consistency with the rest of the codebase / that hyphen has surprisingly caused me some issues.

Reply:
anything that is popped and consumed (and never mentioned again) inside a `from_params` is sort of intrinsically undocumented. my bias would be to just avoid doing that.

e.g. in #1635 since `no_grad_regexes` is a trainer param, it should be a constructor parameter of `Trainer` and turning those gradients off should happen in the Trainer constructor, not right before you call the Trainer constructor.

similarly here (although I haven't wrapped my head fully around this code yet), if there are parameters that are relevant to how the optimizer is instantiated, they should be constructor parameters of `Optimizer` (which isn't really a base class, but if we really need so much logic in it it probably should be).

this is not just an issue of documentation, IMO we should aim not to have functionality in the library that's only available when constructing things `from_params` but not when constructing them directly

Reply:
Yeah, that's a good point, that we shouldn't have functionality that's only accessible through `from_params`.  The trouble is that there are a bunch of things that interact in the `Trainer` that are best constructed outside of it, not inside.  Maybe those things belong in the `Train` command...?  This one needs some more thought, I think.

Reply:
I assume this is addressed by the last commit moving this functionality into `learning_rate_schedulers.py` (https://github.com/allenai/allennlp/pull/1636/commits/a81040279d701163d50f8798631c1faec44f7c16).